commit 478718c68395ea71a8d0ccf6db1107a89b5d73fb
Author:     Yuhong Wen <yh_wen@hotmail.com>
AuthorDate: Fri May 13 20:19:20 2022 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Fri May 13 20:19:20 2022 -0400

    Heartbeat jobs (#524)
    
    * Added running jobs info to the heartbeat.
    
    * Added unit tests for HeartBeat server/client sync up.
    
    * Resolved the conflict with dev-2.1
    
    * renamed a variable.
    
    * Enhanced to have the client abort the extra jobs based on the Hearbeat call response.
    
    * code style change.
    
    * Removed the no used json import.
    
    * Code style change.
---
 nvflare/private/fed/server/job_runner.py | 27 +++++++++++++++++++++------
 1 file changed, 21 insertions(+), 6 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index eea3d6f7..b6961d8b 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -151,13 +151,9 @@ class JobRunner(FLComponent):
         engine = fl_ctx.get_engine()
         run_process = engine.run_processes.get(run_number)
         if run_process:
-            admin_server = engine.server.admin_server
-
             client_sites = run_process.get(RunProcessKey.PARTICIPANTS)
-            message = Message(topic=TrainingTopic.ABORT, body="")
-            message.set_header(RequestHeader.RUN_NUM, str(run_number))
-            self.log_debug(fl_ctx, f"Send stop command to the site for run:{run_number}")
-            replies = self._send_to_clients(admin_server, client_sites, engine, message)
+
+            replies = self.abort_client_run(engine, run_number, client_sites, fl_ctx)
             if not replies:
                 self.log_error(fl_ctx, f"Failed to send abort command to clients for run_{run_number}")
 
@@ -165,6 +161,25 @@ class JobRunner(FLComponent):
             if err:
                 self.log_error(fl_ctx, f"Failed to abort the server for run_.{run_number}")
 
+    def abort_client_run(self, engine, run_number, client_sites, fl_ctx):
+        """Send the abort run command to the clients
+
+        Args:
+            engine: Server Engine
+            run_number: run_number
+            client_sites: Clients to be aborted
+            fl_ctx: FLContext
+
+        Returns:
+
+        """
+        admin_server = engine.server.admin_server
+        message = Message(topic=TrainingTopic.ABORT, body="")
+        message.set_header(RequestHeader.RUN_NUM, str(run_number))
+        self.log_debug(fl_ctx, f"Send stop command to the site for run:{run_number}")
+        replies = self._send_to_clients(admin_server, client_sites, engine, message)
+        return replies
+
     def _delete_run(self, run_number, sites: dict, fl_ctx: FLContext):
         """Delete the run workspace
 
